jsapp/js/users/userExistence.store.ts

```diff
@@ -0,0 +1,67 @@
+import {fetchGet} from 'js/api';
+import {handleApiFail} from 'js/utils';
+import type {FailResponse} from 'js/dataInterface';
+
+export interface UserExistenceStoreData {
+  [username: string]: boolean;
+}
+
+export interface UserResponse {
+  url: string;
+  username: string;
+  date_joined: string;
+  public_collection_subscribers_count: number;
+  public_collections_count: number;
+}
+
+const USERS_USER_ENDPOINT = '/api/v2/users/<username>/';
+
+/**
+ * A store for checking if given user exists.
+ */
+class UserExistenceStore {
+  /** Stores a boolean for known/checked usernames. */
+  private checkedUsers: UserExistenceStoreData = {};
+  private lastCheckDate: Date = new Date();
+
+  /** Resolves with either memoized value or API call response. */
+  async checkUsername(username: string): Promise<boolean> {
+    // If few minutes passed since the last check, we want to wipe the memoized
+    // values, since there is a chance a user might be created in the meantime
+    // and we don't want to risk storing untrue `false` that will never be fixed
+    // (unless user reloads the browser).
+    const now = new Date();
+    const diffMs = (now.valueOf() - this.lastCheckDate.valueOf());
+    const diffMins = Math.round(((diffMs % 86400000) % 3600000) / 60000);
+    if (diffMins >= 3) {
```

Comment:
What's all this conversion about?

```ts
const diffMins = Math.round(((diffMs % 86400000) % 3600000) / 60000);
if (diffMins >= 3) {
```

I have some problems with this code.

- It's hard to read.
- Rounding minutes before comparing `(diffMins >= 3)` makes 2½ minutes the limit instead of 3.
- The modulo `((…) % 3600000)` creates a brief window every hour where cache invalidation doesn't happen. 
- `% 86400000` is redundant with `% 3600000` (they divide evenly).

Suggestion:

- Compare the epoch time in milliseconds.

```ts
const limitMs = 3 * 60 * 1000; // 3 minutes
if (diffMs >= limitMs) {  /* invalidate cache */ }
```

```suggestion
    const limitMs = 3 * 60 * 1000; // 3 minutes
    if (diffMs >= limitMs) {
```

Reply:
A lesson to not trust trusted stack overflow responses 😉